### DISCUSSION TITLE
Difference between sktime SignatureTransformer and esig stream2sig

### DISCUSSION THREAD

**Discussion by wjung21** [alan-turing-institute/sktime#1534] [Q&A] [answered]
Hi, 
I've been working with the esig library, but the versatility of sktime in the signature transformation just brought me here.

I have a question about the SignatureTransformer.fit_transform method. 

```python
from sktime.transformations.panel.signature_based import SignatureTransformer
from sktime.datatypes._panel._convert import from_3d_numpy_to_nested
from esig.tosig import stream2sig

X_example = np.random.rand(1,2,10)
X_nested = from_3d_numpy_to_nested(X_example)

ST = SignatureTransformer(augmentation_list = ('addtime'), window_name = 'global', depth = 2)
y_from_sktime = ST.fit_transform(X_nested)

X = X_example[0]
time = np.expand_dims(np.linspace(0,1,X.shape[1]), axis = 0)
X_withtime = np.concatenate((X,time),0).transpose()
y_from_esig = stream2sig(X_withtime, 2)
print(y_from_esig)
print(y_from_sktime)
```
I thought y_from_esig and y_from_sktime should be the same, but they were different! First, their lengths were different: len(y_from_esig) = 13 but y_from_sktime has 12 columns. 
Also, the values were slightly different. For example, S(1) and S(2) of each signature feature are the same, but features in higher depth are different. 

So, I wonder if there's an implicit difference between the esig computation and the sktime module, or if there is some default internal transformation in sktime.

Any comments?

**Comment by fkiraly** [comment]
`sktime` only provides the interface, @jambo6 who implemented the algorithm interface should be able to comment on this?

Further, note there is currently a known but in that the transformer does not comply with the transformer template #1369, which may or may not be related.

This should be addressed in the transformer refactor #1365 once merged.

**Reply by fkiraly** [reply]
@jambo6, can you kindly comment?

**Comment by jambo6** [comment, marked as answer]
The signature always has first value `=1`. We choose to ignore this in the sktime implementation since it will never add any value to predictive models.

The higher depth differences are coming in because your columns are ordered differently. The sktime implementation always puts time first, so if you check
```
X_withtime = np.concatenate((time, X), 0).transpose()
```
then you will get the same result. 

Note that the values returned in your examples will simply be some permutation of one another. The same values exist, but just in a different order. 

Very sorry for the lateness of this reply!
